.github/workflows/golden.yml

```diff
@@ -0,0 +1,154 @@
+name: golden
+on:
+  push:
+    branches: [main]
+    tags:
+      - "v[0-9]+.[0-9]+.[0-9]+*"
+    paths-ignore:
+      - "**/README.md"
+  merge_group:
+  pull_request:
+    paths-ignore:
+      - "**/README.md"
+env:
+  # Make sure to exit early if cache segment download times out after 2 minutes.
+  # We limit cache download as a whole to 5 minutes.
+  SEGMENT_DOWNLOAD_TIMEOUT_MINS: 2
+
+# Do not cancel this workflow on main. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16616
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+permissions: read-all
+
+jobs:
+  build-dev:
```

Comment:
Can you use one job to build the binaries and upload the artifacts, then the other jobs just depend on that job, download the artifacts, and do their own functionality on top?

I see the `if` conditions are different between jobs, but there should still be some way to make it work properly.

Reply:
I don't think it would actually save time since the jobs are run in parallel, but it would save on some compute not needing to duplicate effort. The main goal for me is just to avoid duplicating workflow code. I'd prefer it to be done here if possible since this is the PR introducing the behavior.

Reply:
I'm not sure I follow, sorry. I copied the build for telemetrygen and replaced with golden. I also don't see where we push the binaries vs the docker image. Can you show me with the telemetrygen build what changes can be done? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/.github/workflows/telemetrygen.yml

Reply:
We can do that but we don't build the same binaries in those jobs. We don't use a collector binary in this build either.